northstar/src/runtime/mount.rs

```diff
@@ -63,101 +63,271 @@ pub enum Error {
-    config: &Config,
-    repo: &Repository,
-) -> Result<Vec<Container>, Error> {
-    info!("Mounting NPKs from {}", repo.dir.display());
-    let mut dir = fs::read_dir(&repo.dir)
-        .await
-        .map_err(|e| Error::Io("Failed to read repository".to_string(), e))?;
-    let mut containers = Vec::new();
-    while let Ok(Some(entry)) = dir.next_entry().await {
-        containers.push(mount_npk(&config, &entry.path(), repo).await?);
-    }
+#[derive(Debug)]
+pub struct MountControl {
+    dm: dm::Dm,
+    lc: LoopControl,
+    run_dir: PathBuf,
+    device_mapper_dev: String,
+}
 
-    Ok(containers)
+async fn mount_point(run_dir: &Path, manifest: &Manifest) -> Result<PathBuf, Error> {
+    let mnt = run_dir
+        .join(&manifest.name)
+        .join(manifest.version.to_string());
+    if !mnt.exists() {
+        info!("Creating mount point {}", mnt.display());
+        fs::create_dir_all(&mnt)
+            .await
+            .map_err(|e| Error::Io(format!("Failed to create directory {}", mnt.display()), e))?;
+    }
+    Ok(mnt)
 }
 
-pub(super) async fn mount_npk(
-    config: &Config,
-    npk: &Path,
-    repo: &Repository,
-) -> Result<Container, Error> {
-    debug!("Mounting {}", npk.display());
-    let use_verity = repo.key.is_some();
+impl MountControl {
+    pub async fn new(config: &Config) -> Result<MountControl, Error> {
+        let lc = LoopControl::open(&config.devices.loop_control, &config.devices.loop_dev)
+            .await
+            .map_err(Error::LoopDevice)?;
+        let dm = dm::Dm::new(&config.devices.device_mapper).map_err(Error::DeviceMapper)?;
+        let run_dir = config.run_dir.clone();
+        let device_mapper_dev = config.devices.device_mapper_dev.clone();
+        Ok(MountControl {
+            lc,
+            dm,
+            run_dir,
+            device_mapper_dev,
+        })
+    }
 
-    let start = time::Instant::now();
+    pub(super) async fn mount_npk_repository(
+        &self,
+        repo: &Repository,
+    ) -> Result<Vec<Container>, Error> {
+        info!("Mounting NPKs from {}", repo.dir.display());
+        let mut dir = fs::read_dir(&repo.dir)
+            .await
+            .map_err(|e| Error::Io("Failed to read repository".to_string(), e))?;
+        let mut containers = Vec::new();
+        while let Ok(Some(entry)) = dir.next_entry().await {
+            containers.push(self.mount_npk(&entry.path(), repo).await?);
+        }
 
-    if let Ok(meta) = metadata(&npk).await {
-        debug!("Mounting NPK with size {}", meta.len());
+        Ok(containers)
     }
 
-    let (manifest, npk_version) = tokio::task::block_in_place(|| {
-        let file = std::fs::File::open(&npk)
-            .map_err(|e| Error::Io(format!("Failed to open NPK at {}", npk.display()), e))?;
-        let mut npk_archive = Npk::new(file).map_err(Error::Npk)?;
+    pub(super) async fn mount_npk(
+        &self,
+        npk: &Path,
+        repo: &Repository,
```

Comment:
The `repo` shouldn't be needed if `npk` is a `Path`.

Reply:
I don't understand :(  `repo` is necessary for the `key` to check with `verity` and to tag the `container` with the repository id.

Reply:
Hm. A repository is basically the key, the name and the dir. You're right to pass the repo instead of it's content split into single arguments.